lib/utils/fs.go

```diff
@@ -241,32 +241,97 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	return fileLock.Unlock, nil
 }
 
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
+	if path == "" {
+		// match behavior from os.RemoveAll
+		return nil
+	}
+	// Match os.RemoveAll protections in not permitting removal of "." directories
+	// This check comes directly from https://cs.opensource.google/go/go/+/refs/tags/go1.21.1:src/os/removeall_at.go;l=24
+	if path == "." || (len(path) >= 2 && path[len(path)-1] == '.' && os.IsPathSeparator(path[len(path)-2])) {
+		return &os.PathError{Op: "RemoveAllSecure", Path: path, Err: syscall.EINVAL} // error type matches os.RemoveAll
+	}
+
+	info, err := os.Lstat(path)
+	switch {
+	case err != nil && os.IsNotExist(err):
+		return nil
+	case err != nil:
+		return trace.ConvertSystemError(err)
+	case !info.IsDir():
+		return removeSecure(path, info)
+	}
+	var removeErrors []error
+	files, err := os.ReadDir(path)
+	if err != nil {
+		// Don't fail fast, allow removal at end to be attempted.
+		removeErrors = []error{err}
+	}
+	// It's possible for a partial file list to be returned even if an error above was returned.
+	for _, f := range files {
+		if err := RemoveAllSecure(filepath.Join(path, f.Name())); err != nil {
+			removeErrors = append(removeErrors, err)
+		}
+	}
+	if err := os.Remove(path); err != nil {
+		removeErrors = append(removeErrors, err)
+	}
+	switch len(removeErrors) {
+	case 1:
+		return trace.ConvertSystemError(removeErrors[0])
+	case 0:
+		return nil
+	default:
+		return trace.NewAggregate(removeErrors...)
+	}
+}
+
 // RemoveSecure attempts to securely delete the file by first overwriting the file with random data three times
 // followed by calling os.Remove(filePath).
 func RemoveSecure(filePath string) error {
-	for i := 0; i < 3; i++ {
-		if err := overwriteFile(filePath); err != nil {
-			return trace.Wrap(err)
-		}
+	info, err := os.Lstat(filePath)
+	if err != nil && os.IsNotExist(err) {
+		return nil
 	}
-	return trace.ConvertSystemError(os.Remove(filePath))
+	// don't fast return on other errors, still allow removeSecure to attempt removal
+	return removeSecure(filePath, info)
 }
 
-func overwriteFile(filePath string) (err error) {
-	f, err := os.OpenFile(filePath, os.O_WRONLY, 0)
-	if err != nil {
-		return trace.ConvertSystemError(err)
+func removeSecure(filePath string, fi os.FileInfo) error {
+	if fi.Mode().Type()&os.ModeSymlink != 0 {
+		return os.Remove(filePath)
+	}
+	f, openErr := os.OpenFile(filePath, os.O_WRONLY, 0)
+	switch {
+	case os.IsNotExist(openErr):
+		return trace.ConvertSystemError(openErr)
+	case openErr != nil:
+		// Attempt delete anyway.
+		return trace.ConvertSystemError(os.Remove(filePath))
 	}
-	defer func() {
-		if closeErr := f.Close(); closeErr != nil {
-			if err == nil {
-				err = trace.ConvertSystemError(closeErr)
-			} else {
-				log.WithError(closeErr).Warningf("Failed to close %v.", f.Name())
+	defer f.Close()
+
+	if runtime.GOOS == "windows" {
```

Comment:
nit: we do have a `WindowsOS` constant somewhere. Not sure if it's worth using, as it may just bloat the import graph.